docs/architecture/adr-031-msg-service.md

```diff
@@ -0,0 +1,216 @@
+For instance, https://github.com/cosmos/cosmos-sdk/issues/7093 proposes a method for
+doing inter-module Ocaps using the `Msg` router. A well-defined return type would
+improve the developer UX for this approach.
+
+## Decision
+
+We decide to add support for using protobuf `service` definitions for `Msg`s.
+
+A `Msg` `service` definition would be written as follows:
+
+```proto
+package cosmos.gov;
+
+service Msg {
+  rpc SubmitProposal(MsgSubmitProposal) returns (MsgSubmitProposalResponse);
+}
+
+// Note that for backwards compatibility this uses MsgSubmitProposal as the request
+// type instead of the more canonical MsgSubmitProposalRequest
+message MsgSubmitProposal {
+  google.protobuf.Any content = 1;
+  string proposer = 2;
+}
+
+message MsgSubmitProposalResponse {
+  uint64 proposal_id;
+}
+```
+
+While this is most commonly used for gRPC, overloading protobuf `service` definitions like this does not violate
+the intent of the [protobuf spec](https://developers.google.com/protocol-buffers/docs/proto3#services) which says:
+> If you don’t want to use gRPC, it’s also possible to use protocol buffers with your own RPC implementation.
+With this approach, we would get an auto-generated `MsgServer` interface:
+
+In addition to clearly specifying return types, this has the benefit of generating client and server code. On the server
+side, this is almost like an automatically generated keeper method and could maybe be used intead of keepers eventually
+(see [\#7093](https://github.com/cosmos/cosmos-sdk/issues/7093)):
+
+```go
+type MsgServer interface {
+  SubmitProposal(context.Context, *MsgSubmitProposal) (*MsgSubmitProposalResponse, error)
+}
+```
+
+On the client side, developers could take advantage of this by creating RPC implementations that encapsulate transaction
+logic. Protobuf libraries that use asynchronous callbacks, like [protobuf.js](https://github.com/protobufjs/protobuf.js#using-services)
+could use this to register callbacks for specific messages even for transactions that include multiple `Msg`s.
+
+For backwards compatibility, existing `Msg` types should be used as the request parameter
+for `service` definitions. Newer `Msg` types which only support `service` definitions
+should use the more canonical `Msg...Request` names.
+
+### Encoding
+
+Currently, we are encoding `Msg`s as `Any` in `Tx`s which involves packing the
+binary-encoded `Msg` with its type URL.
+
+The type URL for `MsgSubmitProposal` is `/cosmos.gov.MsgSubmitProposal`. 
+
+The fully-qualified name for the `SubmitProposal` service method above is
+`/cosmos.gov.Msg/SubmitProposal` which is varies by a single `/` character.
+
+In order to encode service methods in transactions, we encode them as `Any`s in
+the same `TxBody.messages` field as other `Msg`s. We simply set `Any.type_url`
+to the full-qualified method name (ex. `/cosmos.gov.Msg/SubmitProposal`) and
+set `Any.value` to the protobuf encoding of the request message
```

Comment:
Is it possible to avoid `Any` here?
I see that this is the way to do it given the `Tx.Msg` type. That being said I already feel that we are abusing it.

Reply:
Well I'm trying not to break anything here as a starting point. We may choose to break later when amino is fully deprecated. Any is already used here so I'm not trying to change that.

Reply:
Also why would you say we're abusing it?

Reply:
It's all clear now, I thought it's not for constructing a transaction. Thanks.